runtimes/rust/lbr-prelude/src/json.rs

```diff
@@ -0,0 +1,327 @@
+                    .map(|val| T::from_json(val.clone()))
+                    .into_iter()
+                    .collect::<Result<HashSet<T>, Error>>()?;
+
+                if set.len() == vec.len() {
+                    Ok(set)
+                } else {
+                    error(&format!(
+                        "Expected the JSON Array to have all unique elements: {:?}",
+                        vec
+                    ))
+                }
+            }
+            _ => error(&format!("Expected a JSON Array but got {:?}", value)),
+        }
+    }
+}
+
+impl<K, V> Json for HashMap<K, V>
+where
+    K: Json + Eq + Hash,
+    V: Json,
+{
+    fn to_json(&self) -> Result<Value, Error> {
+        let values = self
+            .iter()
+            .map(|(key, val)| Ok(Value::Array(vec![key.to_json()?, val.to_json()?])))
+            .into_iter()
+            .collect::<Result<Vec<Value>, Error>>()?;
+
+        Ok(Value::Array(values))
+    }
+
+    fn from_json(value: Value) -> Result<Self, Error> {
+        match value {
+            Value::Array(vec) => {
+                let set = vec
+                    .iter()
+                    .map(|kv_tuple| <(K, V)>::from_json(kv_tuple.clone()))
+                    .into_iter()
+                    .collect::<Result<HashMap<K, V>, Error>>()?;
+
+                if set.len() == vec.len() {
+                    Ok(set)
+                } else {
+                    error(&format!(
+                        "Expected the JSON Array to have all unique elements: {:?}",
+                        vec
+                    ))
+                }
+            }
+            _ => error(&format!("Expected a JSON Array but got {:?}", value)),
+        }
+    }
+}
+
+impl Json for () {
+    fn to_json(&self) -> Result<Value, Error> {
+        Ok(Value::Null)
+    }
+
+    fn from_json(value: Value) -> Result<Self, Error> {
+        match value {
+            Value::Null => Ok(()),
+            _ => error(&format!("Expected a JSON Null but got {:?}", value)),
+        }
+    }
+}
+
+impl Json for Value {
+    fn to_json(&self) -> Result<Value, Error> {
+        Ok(self.clone())
+    }
+
+    fn from_json(value: Value) -> Result<Self, Error> {
+        Ok(value)
+    }
+}
+
+impl<A, B> Json for (A, B)
+where
+    A: Json,
+    B: Json,
+{
+    fn to_json(&self) -> Result<Value, Error> {
+        Ok(Value::Array(vec![self.0.to_json()?, self.1.to_json()?]))
+    }
+
+    fn from_json(value: Value) -> Result<Self, Error> {
+        match value {
+            Value::Array(vec) => match &vec[..] {
+                [a, b] => Ok((A::from_json(a.clone())?, B::from_json(b.clone())?)),
+                _ => error("Expected a JSON Array with 2 fields"),
+            },
+            _ => error("Expected a JSON Array with 2 fields"),
+        }
+    }
+}
+
+/// Create a JSON serialisation error with a custom message
+fn error<T>(msg: &str) -> Result<T, Error> {
+    Err(ser::Error::custom(msg))
+}
```

Comment:
I know I didn't do that in the Hs/Purs counterparts, but I like the pattern of having an `errors.rs` module where the errors used by all (sub)modules. There we can enumerate all the errors in some convenient form and have a single place to control how it;s printed.

For example:
```rust
#[derive(thiserror::Error, Debug)]
enum Error {
  #[error("Expected a JSON type: {1}, but got a JSON type: {0}")]
  UnexpectedJsonType {got: JsonType, wanted: JsonType},
  #[error("Expected a field name in a JSON Object: {1}, but got fields named {0}")]  
  UnexpectedFieldName {got: Vec<String>, wanted: String},
  #[error("Expected a JSON Array with {1} elements, but got {0} elements")]  
  UnexpectedArrayLength {got: int, wanted: int},
  #[error("Some internal error happened: {0}")]
  InternalError(String),
}
```

Reply:
I want to introduce that in our Hs/Purs base, Rust folks have good simple conventions that improve everyone's lives so why not cross polinate.

Reply:
Sounds like a good idea.